compiler/rustc_mir_build/src/thir/cx/expr.rs

```diff
@@ -351,19 +351,34 @@ impl<'tcx> Cx<'tcx> {
                             });
                         }
                     }
-                    let adt_data =
-                        if let hir::ExprKind::Path(hir::QPath::Resolved(_, ref path)) = fun.kind {
+                    let adt_data = if let hir::ExprKind::Path(qpath) = fun.kind {
```

Comment:
```suggestion
                    let adt_data = if let hir::ExprKind::Path(qpath) = fun.kind
                        && let Some(adt_def) = expr_ty.ty_adt_def()
                    {
```
To avoid having to do it in all branches.